src/main/java/org/opensearch/ad/AnomalyDetectorJobRunner.java

```diff
@@ -1,698 +1,697 @@
+    private Settings settings;
+    private int maxRetryForEndRunException;
+    private Client client;
+    private ThreadPool threadPool;
+    private AnomalyIndexHandler<AnomalyResult> anomalyResultHandler;
+    private ConcurrentHashMap<String, Integer> detectorEndRunExceptionCount;
+    private AnomalyDetectionIndices anomalyDetectionIndices;
+    private DiscoveryNodeFilterer nodeFilter;
+    private ADTaskManager adTaskManager;
+
+    public static AnomalyDetectorJobRunner getJobRunnerInstance() {
+        if (INSTANCE != null) {
+            return INSTANCE;
+        }
+        synchronized (AnomalyDetectorJobRunner.class) {
+            if (INSTANCE != null) {
+                return INSTANCE;
+            }
+            INSTANCE = new AnomalyDetectorJobRunner();
+            return INSTANCE;
+        }
+    }
+
+    private AnomalyDetectorJobRunner() {
+        // Singleton class, use getJobRunnerInstance method instead of constructor
+        this.detectorEndRunExceptionCount = new ConcurrentHashMap<>();
+    }
+
+    public void setClient(Client client) {
+        this.client = client;
+    }
+
+    public void setThreadPool(ThreadPool threadPool) {
+        this.threadPool = threadPool;
+    }
+
+    public void setAnomalyResultHandler(AnomalyIndexHandler<AnomalyResult> anomalyResultHandler) {
+        this.anomalyResultHandler = anomalyResultHandler;
+    }
+
+    public void setSettings(Settings settings) {
+        this.settings = settings;
+        this.maxRetryForEndRunException = AnomalyDetectorSettings.MAX_RETRY_FOR_END_RUN_EXCEPTION.get(settings);
+    }
+
+    public void setAdTaskManager(ADTaskManager adTaskManager) {
+        this.adTaskManager = adTaskManager;
+    }
+
+    public void setAnomalyDetectionIndices(AnomalyDetectionIndices anomalyDetectionIndices) {
+        this.anomalyDetectionIndices = anomalyDetectionIndices;
+    }
+
+    public void setNodeFilter(DiscoveryNodeFilterer nodeFilter) {
+        this.nodeFilter = nodeFilter;
+    }
+
+    @Override
+    public void runJob(ScheduledJobParameter scheduledJobParameter, JobExecutionContext context) {
+        String detectorId = scheduledJobParameter.getName();
+        log.info("Start to run AD job {}", detectorId);
+        adTaskManager.refreshRealtimeJobRunTime(detectorId);
+        if (!(scheduledJobParameter instanceof AnomalyDetectorJob)) {
+            throw new IllegalArgumentException(
+                "Job parameter is not instance of AnomalyDetectorJob, type: " + scheduledJobParameter.getClass().getCanonicalName()
+            );
+        }
+        AnomalyDetectorJob jobParameter = (AnomalyDetectorJob) scheduledJobParameter;
+        Instant executionStartTime = Instant.now();
+        IntervalSchedule schedule = (IntervalSchedule) jobParameter.getSchedule();
+        Instant detectionStartTime = executionStartTime.minus(schedule.getInterval(), schedule.getUnit());
+
+        final LockService lockService = context.getLockService();
```

Comment:
We'll need to modify this workflow and avoid pulling the lockService from the provided JobExecutionContext. The writeable JobExecutionContext sent from Job Scheduler will set the lockService as null since this is not serializable, however we shall be handling this replacement as part of work items 5 and 6 for the job scheduler communication mechanism. 

For now, it is fine to leave this as is, but add //TODOs to mark this implementation for replacement to invoke the REST api's on JS to acquire/release locks

Reply:
ACK